test/add/common/add_test.dart

```diff
@@ -0,0 +1,598 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:test/test.dart';
+
+import 'package:pub/src/exit_codes.dart' as exit_codes;
+
+import '../../descriptor.dart' as d;
+import '../../test_pub.dart';
+
+void main() {
+  test('URL encodes the package name', () async {
+    await serveNoPackages();
+
+    await d.appDir({}).create();
+
+    await pubAdd(
+        args: ['bad name!:1.2.3'],
+        error: allOf([
+          contains(
+              "Because myapp depends on bad name! any which doesn't exist (could "
+              'not find package bad name! at http://localhost:'),
+          contains('), version solving failed.')
+        ]),
+        exitCode: exit_codes.UNAVAILABLE);
+
+    await d.appDir({}).validate();
+  });
+
+  group('normally', () {
+    test('adds a package from a pub server', () async {
+      await servePackages((builder) => builder.serve('foo', '1.2.3'));
+
+      await d.appDir({}).create();
+
+      await pubAdd(args: ['foo:1.2.3']);
+
+      await d.cacheDir({'foo': '1.2.3'}).validate();
+      await d.appPackagesFile({'foo': '1.2.3'}).validate();
+      await d.appDir({'foo': '1.2.3'}).validate();
+    });
+
+    test('dry run does not actually add the package or modify the pubspec',
+        () async {
+      await servePackages((builder) => builder.serve('foo', '1.2.3'));
+
+      await d.appDir({}).create();
+
+      await pubAdd(
+          args: ['foo:1.2.3', '--dry-run'],
+          output: allOf([
+            contains('Would change 1 dependency'),
+            contains('+ foo 1.2.3')
+          ]));
+
+      await d.appDir({}).validate();
+    });
+
+    test(
+        'adds a package from a pub server even when dependencies key does not exist',
+        () async {
+      await servePackages((builder) => builder.serve('foo', '1.2.3'));
+
+      await d.dir(appPath, [
+        d.pubspec({'name': 'myapp'})
+      ]).create();
+
+      await pubAdd(args: ['foo:1.2.3']);
+
+      await d.cacheDir({'foo': '1.2.3'}).validate();
+      await d.appPackagesFile({'foo': '1.2.3'}).validate();
+      await d.appDir({'foo': '1.2.3'}).validate();
+    });
+
+    group('overrides existing version constraint if package exists', () {
+      test('if package is added without a version constraint', () async {
+        await servePackages((builder) {
+          builder.serve('foo', '1.2.3');
+          builder.serve('foo', '1.2.2');
+        });
+
+        await d.appDir({'foo': '1.2.2'}).create();
+
+        await pubAdd(args: ['foo']);
+
+        await d.cacheDir({'foo': '1.2.3'}).validate();
+        await d.appPackagesFile({'foo': '1.2.3'}).validate();
+        await d.appDir({'foo': '^1.2.3'}).validate();
```

Comment:
Let's not overwrite in this case... let's instead print a message to go use  `pub upgrade --bump foo`.

Reply:
there is a case where... we have a dependency on `foo: ^1.2.3` 
and the current version is `1.2.3` and the latest available version is `1.2.4`, in this case `pub add foo` should print that we already have a dependency on `foo`, this can be upgraded with `pub upgrade foo`.

Reply:
I made a temporary message to use `pub upgrade foo`, with a TODO to revise it later!